add megatron training doc (#4159)
ROCm/ROCm: docs/how-to/rocm-for-ai/train-a-model.rst

   :description: How to use ROCm for AI

>    :keywords: ROCm, AI, LLM, train, fine-tune, FSDP, DeepSpeed, LLaMA, tutorial

Reviewer: Have the keywords been updated?
The description metadata needs updating

**************************************
Training a model with ROCm Megatron-LM
**************************************

.. _amd-megatron-lm:

The ROCm Megatron-LM framework is a specialized fork of the robust Megatron-LM, designed to
enable efficient training of large-scale language models on AMD GPUs. By leveraging AMD Instinct™ MI300X
accelerators, AMD Megatron-LM delivers enhanced scalability, performance, and resource utilization for AI
workloads. It is purpose-built to :ref:`support models <amd-megatron-lm-model-support>`
like Llama 2, Llama 3, and Llama 3.1, enabling developers to train next-generation AI models with greater

> efficiency. See the GitHub repository at `<https://github.com/ROCm/Megatron-LM>`__.

Author: Yes, readers will clone https://github.com/ROCm/Megatron-LM in subsequent instructions

For ease of use, AMD provides a ready-to-use Docker image for MI300X accelerators containing essential

> components including PyTorch, PyTorch Lightning, ROCm libraries, and Megatron-LM utilities. It contains the

Reviewer: ```suggestion
components, including PyTorch, PyTorch Lightning, ROCm libraries, and Megatron-LM utilities. It contains the
```

> following software to accelerate training workloads:

Reviewer: software? or applications?

Author: I prefer software in this instance

+--------------------------+--------------------------------+
| PyTorch Lightning        | 2.4.0                          |
+--------------------------+--------------------------------+
| Megatron Core            | 0.9.0                          |
+--------------------------+--------------------------------+
| Transformer Engine       | 1.5.0                          |
+--------------------------+--------------------------------+
| Flash Attention          | v2.6                           |
+--------------------------+--------------------------------+
| Transformers             | 4.44.0                         |
+--------------------------+--------------------------------+

Supported features and models
=============================

Megatron-LM provides the following key features to train large language models efficiently:

- Transformer Engine (TE)

- APEX

- GEMM tuning

- Torch.compile

- 3D parallelism: TP + SP + CP

- Distributed optimizer

- Flash Attention (FA) 2

- Fused kernels

- Pre-training

.. _amd-megatron-lm-model-support:

The following models are pre-optimized for performance on the AMD Instinct MI300X accelerator.

> * Llama 2 7B

Reviewer: Perhaps the Llama should be using Meta on the first occurrence. 
https://ai.meta.com/blog/meta-llama-3/

Author: Done @ first occurrence

* Llama 2 70B

* Llama 3 8B

* Llama 3 70B

* Llama 3.1 8B

* Llama 3.1 70B

Prerequisite system validation steps
====================================

Complete the following system validation and optimization steps to set up your system before starting training.

Disable NUMA auto-balancing
---------------------------

> Generally, application performance can benefit from disabling NUMA auto-balancing; however, there are
> certain types of workloads where doing so might be detrimental to performance.

Reviewer: ```suggestion
Generally, application performance can benefit from disabling NUMA auto-balancing. However, it might be detrimental to performance with certain workloads.
```

.. code-block:: shell

   sudo sh -c 'echo 0 > /proc/sys/kernel/numa_balancing'

See :ref:`mi300x-disable-numa` for more information.

Hardware verification with ROCm
-------------------------------

Use the command ``rocm-smi --setperfdeterminism 1900`` to set the max clock speed up to 1900 MHz
instead of the default 2100 MHz. This can reduce the chance of a PCC event lowering the attainable
GPU clocks. This setting will not be required for new IFWI releases with the production PRC feature.
You can restore this setting to its default value with the ``rocm-smi -r`` command.

Run the command:

.. code-block:: shell

   rocm-smi --setperfdeterminism 1900

See :ref:`mi300x-hardware-verification-with-rocm` for more information.

RCCL Bandwidth Test
-------------------

ROCm Collective Communications Library (RCCL) is a standalone library of standard collective communication
routines for GPUs. See the :doc:`RCCL documentation <rccl:index>` for more information. Before starting
pre-training, running a RCCL bandwidth test helps ensure that the multi-GPU or multi-node setup is optimized
for efficient distributed training.

Running the RCCL bandwidth test helps verify that:

> - The GPUs can communicate across nodes or within a single node.
> 
> - The interconnect (such as InfiniBand, Ethernet, or Infinite fabric) is functioning as expected and
>   provides adequate bandwidth for communication.
> 
> - There are no hardware setup or cabling issues that could affect the
>   communication between GPUs.
> 

Reviewer: ```suggestion
- The GPUs can communicate across nodes or within a single node.

- The interconnect (such as InfiniBand, Ethernet, or Infinite fabric) functions as expected and
  provides adequate bandwidth for communication.

- No hardware setup or cabling issues could affect the communication between GPUs.

```
